source/opt/scalar_analysis_nodes.h

```diff
@@ -0,0 +1,284 @@
+// is an expression with a loop variant as one of its terms, such as an
+// induction variable. The actual value of a recurrent expression is coefficent_
+// * iteration + offset_, hence an induction variable i=0, i++ becomes a
+// recurrent expression with an offset of zero and a coefficient of one.
+class SERecurrentNode : public SENode {
+ public:
+  SERecurrentNode(opt::ScalarEvolutionAnalysis* parent_analysis,
+                  const ir::Loop* loop)
+      : SENode(parent_analysis), loop_(loop) {}
+
+  SENodeType GetType() const final { return RecurrentExpr; }
+
+  inline void AddCoefficient(SENode* child) {
+    coefficient_ = child;
+    SENode::AddChild(child);
+  }
+
+  inline void AddOffset(SENode* child) {
+    offset_ = child;
+    SENode::AddChild(child);
+  }
+
+  inline const SENode* GetCoefficient() const { return coefficient_; }
+  inline SENode* GetCoefficient() { return coefficient_; }
+
+  inline const SENode* GetOffset() const { return offset_; }
+  inline SENode* GetOffset() { return offset_; }
+
+  // Return the loop which this recurrent expression is recurring within.
+  const ir::Loop* GetLoop() const { return loop_; }
+
+  SERecurrentNode* AsSERecurrentNode() override { return this; }
+  const SERecurrentNode* AsSERecurrentNode() const override { return this; }
+
+ private:
+  SENode* coefficient_;
+  SENode* offset_;
+  const ir::Loop* loop_;
+};
+
+// A node representing an addition operation between child nodes.
+class SEAddNode : public SENode {
+ public:
+  explicit SEAddNode(opt::ScalarEvolutionAnalysis* parent_analysis)
+      : SENode(parent_analysis) {}
+
+  SENodeType GetType() const final { return Add; }
+
+  SEAddNode* AsSEAddNode() override { return this; }
+  const SEAddNode* AsSEAddNode() const override { return this; }
+};
+
+// A node representing a multiply operation between child nodes.
+class SEMultiplyNode : public SENode {
+ public:
+  explicit SEMultiplyNode(opt::ScalarEvolutionAnalysis* parent_analysis)
+      : SENode(parent_analysis) {}
+
+  SENodeType GetType() const final { return Multiply; }
+
+  SEMultiplyNode* AsSEMultiplyNode() override { return this; }
+  const SEMultiplyNode* AsSEMultiplyNode() const override { return this; }
+};
+
+// A node representing a unary negative operation.
+class SENegative : public SENode {
+ public:
+  explicit SENegative(opt::ScalarEvolutionAnalysis* parent_analysis)
+      : SENode(parent_analysis) {}
+
+  SENodeType GetType() const final { return Negative; }
+
+  SENegative* AsSENegative() override { return this; }
+  const SENegative* AsSENegative() const override { return this; }
+};
+
+// A node representing a value which we do not know the value of, such as a load
+// instruction.
+class SEValueUnknown : public SENode {
+ public:
+  // SEValueUnknowns must come from an instruction |unique_id| is the unique id
+  // of that instruction. This is so we can compare value unknowns and have a
+  // unique value unknown for each instruction.
```

Comment:
I think I would prefer the result id to be used.  Is there a reason it cannot be used?

Reply:
No reason, updated to use result now